File: csharp/utils/Log.cs
namespace adventofcode.utils;

using System;

public enum LogLevel
{
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3,
    None = 4
}

public static class Log
{
    private static LogLevel _minLevel = LogLevel.Info;

    public static void SetLevel(LogLevel level) => _minLevel = level;

    public static void Debug(string message) => _log(LogLevel.Debug, message);
    public static void Info(string message) => _log(LogLevel.Info, message);
    public static void Warn(string message) => _log(LogLevel.Warn, message);
    public static void Error(string message) => _log(LogLevel.Error, message);

    private static void _log(LogLevel level, string message)
    {
        if (level < _minLevel || _minLevel == LogLevel.None) return;

        var levelName = level.ToString().ToUpperInvariant();
        Console.WriteLine($"[{levelName}] {message}");
    }
}

File: java/src/main/java/adventofcode/util/Log.java
package adventofcode.util;

public class Log {
    public enum Level {
        DEBUG(10),
        INFO(20),
        WARN(30),
        ERROR(40);

        private final int value;

        Level(int value) {
            this.value = value;
        }

        public int value() {
            return value;
        }
    }

    private static Level configuredLevel = Level.INFO;

    public static void setLogLevel(Level level) {
        configuredLevel = level;
    }

    public static void debug(String msg, Object... args) {
        doLog(Level.DEBUG, msg, args);
    }

    public static void info(String msg, Object... args) {
        doLog(Level.INFO, msg, args);
    }

    public static void warn(String msg, Object... args) {
        doLog(Level.WARN, msg, args);
    }

    public static void error(String msg, Object... args) {
        doLog(Level.ERROR, msg, args);
    }

    private static void doLog(Level level, String msg, Object[] args) {
        if (configuredLevel.value() <= level.value()) {
            System.out.println(level.name() + " | " + String.format(msg, args));
        }
    }
}
